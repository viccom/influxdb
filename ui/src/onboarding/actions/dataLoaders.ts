// Libraries
import _ from 'lodash'

// Apis
import {
  writeLineProtocol,
  createTelegrafConfig,
  getTelegrafConfigs,
  updateTelegrafConfig,
  createScraperTarget,
  updateScraperTarget,
} from 'src/onboarding/apis/index'

// Utils
import {createNewPlugin} from 'src/onboarding/utils/pluginConfigs'

// Constants
import {
  pluginsByBundle,
  telegrafPluginsInfo,
} from 'src/onboarding/constants/pluginConfigs'

// Types
import {
  TelegrafPlugin,
  TelegrafPluginName,
  DataLoaderType,
  LineProtocolTab,
  Plugin,
  BundleName,
  ConfigurationState,
} from 'src/types/v2/dataLoaders'
import {AppState} from 'src/types/v2'
import {RemoteDataState} from 'src/types'
import {
  WritePrecision,
  TelegrafRequest,
  TelegrafPluginOutputInfluxDBV2,
} from 'src/api'
import {Dispatch} from 'redux'

type GetState = () => AppState

const DEFAULT_COLLECTION_INTERVAL = 10000

export type Action =
  | SetDataLoadersType
  | SetTelegrafConfigID
  | UpdateTelegrafPluginConfig
  | AddConfigValue
  | RemoveConfigValue
  | SetActiveTelegrafPlugin
  | SetLineProtocolBody
  | SetActiveLPTab
  | SetLPStatus
  | SetPrecision
  | UpdateTelegrafPlugin
  | AddPluginBundle
  | AddTelegrafPlugins
  | RemoveBundlePlugins
  | RemovePluginBundle
  | SetPluginConfiguration
  | SetConfigArrayValue
  | SetScraperTargetBucket
  | SetScraperTargetURL
  | SetScraperTargetID
  | ClearDataLoaders

interface SetDataLoadersType {
  type: 'SET_DATA_LOADERS_TYPE'
  payload: {type: DataLoaderType}
}

export const setDataLoadersType = (
  type: DataLoaderType
): SetDataLoadersType => ({
  type: 'SET_DATA_LOADERS_TYPE',
  payload: {type},
})

interface ClearDataLoaders {
  type: 'CLEAR_DATA_LOADERS'
}

export const clearDataLoaders = (): ClearDataLoaders => ({
  type: 'CLEAR_DATA_LOADERS',
})

interface UpdateTelegrafPluginConfig {
  type: 'UPDATE_TELEGRAF_PLUGIN_CONFIG'
  payload: {name: string; field: string; value: string}
}

export const updateTelegrafPluginConfig = (
  name: string,
  field: string,
  value: string
): UpdateTelegrafPluginConfig => ({
  type: 'UPDATE_TELEGRAF_PLUGIN_CONFIG',
  payload: {name, field, value},
})

interface UpdateTelegrafPlugin {
  type: 'UPDATE_TELEGRAF_PLUGIN'
  payload: {plugin: Plugin}
}

export const updateTelegrafPlugin = (plugin: Plugin): UpdateTelegrafPlugin => ({
  type: 'UPDATE_TELEGRAF_PLUGIN',
  payload: {plugin},
})

interface AddConfigValue {
  type: 'ADD_TELEGRAF_PLUGIN_CONFIG_VALUE'
  payload: {
    pluginName: string
    fieldName: string
    value: string
  }
}

export const addConfigValue = (
  pluginName: string,
  fieldName: string,
  value: string
): AddConfigValue => ({
  type: 'ADD_TELEGRAF_PLUGIN_CONFIG_VALUE',
  payload: {pluginName, fieldName, value},
})

interface RemoveConfigValue {
  type: 'REMOVE_TELEGRAF_PLUGIN_CONFIG_VALUE'
  payload: {
    pluginName: string
    fieldName: string
    value: string
  }
}

export const removeConfigValue = (
  pluginName: string,
  fieldName: string,
  value: string
): RemoveConfigValue => ({
  type: 'REMOVE_TELEGRAF_PLUGIN_CONFIG_VALUE',
  payload: {pluginName, fieldName, value},
})

interface SetConfigArrayValue {
  type: 'SET_TELEGRAF_PLUGIN_CONFIG_VALUE'
  payload: {
    pluginName: TelegrafPluginName
    field: string
    valueIndex: number
    value: string
  }
}

export const setConfigArrayValue = (
  pluginName: TelegrafPluginName,
  field: string,
  valueIndex: number,
  value: string
): SetConfigArrayValue => ({
  type: 'SET_TELEGRAF_PLUGIN_CONFIG_VALUE',
  payload: {pluginName, field, valueIndex, value},
})

interface SetTelegrafConfigID {
  type: 'SET_TELEGRAF_CONFIG_ID'
  payload: {id: string}
}

export const setTelegrafConfigID = (id: string): SetTelegrafConfigID => ({
  type: 'SET_TELEGRAF_CONFIG_ID',
  payload: {id},
})

interface AddPluginBundle {
  type: 'ADD_PLUGIN_BUNDLE'
  payload: {bundle: BundleName}
}

export const addPluginBundle = (bundle: BundleName): AddPluginBundle => ({
  type: 'ADD_PLUGIN_BUNDLE',
  payload: {bundle},
})

interface RemovePluginBundle {
  type: 'REMOVE_PLUGIN_BUNDLE'
  payload: {bundle: BundleName}
}

export const removePluginBundle = (bundle: BundleName): RemovePluginBundle => ({
  type: 'REMOVE_PLUGIN_BUNDLE',
  payload: {bundle},
})
interface AddTelegrafPlugins {
  type: 'ADD_TELEGRAF_PLUGINS'
  payload: {telegrafPlugins: TelegrafPlugin[]}
}

export const addTelegrafPlugins = (
  telegrafPlugins: TelegrafPlugin[]
): AddTelegrafPlugins => ({
  type: 'ADD_TELEGRAF_PLUGINS',
  payload: {telegrafPlugins},
})

interface RemoveBundlePlugins {
  type: 'REMOVE_BUNDLE_PLUGINS'
  payload: {bundle: BundleName}
}

export const removeBundlePlugins = (
  bundle: BundleName
): RemoveBundlePlugins => ({
  type: 'REMOVE_BUNDLE_PLUGINS',
  payload: {bundle},
})

interface SetScraperTargetBucket {
  type: 'SET_SCRAPER_TARGET_BUCKET'
  payload: {bucket: string}
}

export const setScraperTargetBucket = (
  bucket: string
): SetScraperTargetBucket => ({
  type: 'SET_SCRAPER_TARGET_BUCKET',
  payload: {bucket},
})

interface SetScraperTargetURL {
  type: 'SET_SCRAPER_TARGET_URL'
  payload: {url: string}
}

export const setScraperTargetURL = (url: string): SetScraperTargetURL => ({
  type: 'SET_SCRAPER_TARGET_URL',
  payload: {url},
})

interface SetScraperTargetID {
  type: 'SET_SCRAPER_TARGET_ID'
  payload: {id: string}
}

export const setScraperTargetID = (id: string): SetScraperTargetID => ({
  type: 'SET_SCRAPER_TARGET_ID',
  payload: {id},
})

export const addPluginBundleWithPlugins = (bundle: BundleName) => dispatch => {
  dispatch(addPluginBundle(bundle))
  const plugins = pluginsByBundle[bundle]
  dispatch(
    addTelegrafPlugins(
      plugins.map(p => {
        const isConfigured = !!telegrafPluginsInfo[p].fields
          ? ConfigurationState.Unconfigured
          : ConfigurationState.Configured

        return {
          name: p,
          active: false,
          configured: isConfigured,
        }
      })
    )
  )
}

export const removePluginBundleWithPlugins = (
  bundle: BundleName
) => dispatch => {
  dispatch(removePluginBundle(bundle))
  dispatch(removeBundlePlugins(bundle))
}

export const createOrUpdateTelegrafConfigAsync = (authToken: string) => async (
  dispatch,
  getState: GetState
) => {
  const {
    dataLoading: {
      dataLoaders: {telegrafPlugins},
      steps: {org, bucket, orgID},
    },
  } = getState()

  const telegrafConfigsFromServer = await getTelegrafConfigs(orgID)

  const influxDB2Out = {
    name: TelegrafPluginOutputInfluxDBV2.NameEnum.InfluxdbV2,
    type: TelegrafPluginOutputInfluxDBV2.TypeEnum.Output,
    config: {
      urls: ['http://127.0.0.1:9999'],
      token: authToken,
      organization: org,
      bucket,
    },
  }

  let plugins: Plugin[] = [influxDB2Out]
  telegrafPlugins.forEach(tp => {
    if (tp.configured === ConfigurationState.Configured) {
      plugins = [...plugins, tp.plugin || createNewPlugin(tp.name)]
    }
  })

  const telegrafRequest: TelegrafRequest = {
    name: 'new config',
    agent: {collectionInterval: DEFAULT_COLLECTION_INTERVAL},
    organizationID: orgID,
    plugins,
  }

  if (telegrafConfigsFromServer.length) {
    const id = _.get(telegrafConfigsFromServer, '0.id', '')

    await updateTelegrafConfig(id, telegrafRequest)
    dispatch(setTelegrafConfigID(id))
    return
  }

  const created = await createTelegrafConfig(telegrafRequest)
  dispatch(setTelegrafConfigID(created.id))
}

interface SetActiveTelegrafPlugin {
  type: 'SET_ACTIVE_TELEGRAF_PLUGIN'
  payload: {telegrafPlugin: string}
}

export const setActiveTelegrafPlugin = (
  telegrafPlugin: string
): SetActiveTelegrafPlugin => ({
  type: 'SET_ACTIVE_TELEGRAF_PLUGIN',
  payload: {telegrafPlugin},
})

interface SetPluginConfiguration {
  type: 'SET_PLUGIN_CONFIGURATION_STATE'
  payload: {telegrafPlugin: TelegrafPluginName}
}

export const setPluginConfiguration = (
  telegrafPlugin: TelegrafPluginName
): SetPluginConfiguration => ({
  type: 'SET_PLUGIN_CONFIGURATION_STATE',
  payload: {telegrafPlugin},
})

interface SetLineProtocolBody {
  type: 'SET_LINE_PROTOCOL_BODY'
  payload: {lineProtocolBody: string}
}

export const setLineProtocolBody = (
  lineProtocolBody: string
): SetLineProtocolBody => ({
  type: 'SET_LINE_PROTOCOL_BODY',
  payload: {lineProtocolBody},
})

interface SetActiveLPTab {
  type: 'SET_ACTIVE_LP_TAB'
  payload: {activeLPTab: LineProtocolTab}
}

export const setActiveLPTab = (
  activeLPTab: LineProtocolTab
): SetActiveLPTab => ({
  type: 'SET_ACTIVE_LP_TAB',
  payload: {activeLPTab},
})

interface SetLPStatus {
  type: 'SET_LP_STATUS'
  payload: {lpStatus: RemoteDataState}
}

export const setLPStatus = (lpStatus: RemoteDataState): SetLPStatus => ({
  type: 'SET_LP_STATUS',
  payload: {lpStatus},
})

interface SetPrecision {
  type: 'SET_PRECISION'
  payload: {precision: WritePrecision}
}

export const setPrecision = (precision: WritePrecision): SetPrecision => ({
  type: 'SET_PRECISION',
  payload: {precision},
})

export const writeLineProtocolAction = (
  org: string,
  bucket: string,
  body: string,
  precision: WritePrecision
) => async dispatch => {
  try {
    dispatch(setLPStatus(RemoteDataState.Loading))
    await writeLineProtocol(org, bucket, body, precision)
    dispatch(setLPStatus(RemoteDataState.Done))
  } catch (error) {
    dispatch(setLPStatus(RemoteDataState.Error))
  }
}

export const saveScraperTarget = () => async (
  dispatch: Dispatch<Action>,
  getState: GetState
) => {
  const {
    onboarding: {bucketID, orgID},
    dataLoading: {
      dataLoaders: {
        scraperTarget: {url, id},
      },
    },
  } = getState()

  try {
    if (id) {
      await updateScraperTarget(id, url, bucketID)
    } else {
      const newTarget = await createScraperTarget(url, orgID, bucketID)
      dispatch(setScraperTargetID(newTarget.id))
    }
  } catch (error) {
    console.error()
  }
}
