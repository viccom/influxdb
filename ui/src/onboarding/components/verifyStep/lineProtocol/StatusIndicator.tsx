// Libraries
import React, {PureComponent} from 'react'
import classnames from 'classnames'

import {SparkleSpinner} from 'src/clockface'

// Types
import {RemoteDataState} from 'src/types'

interface Props {
  status: RemoteDataState
  onClickRetry: () => void
}

class StatusIndicator extends PureComponent<Props> {
  public render() {
    const {status} = this.props
    return (
      <>
        <div className={'wizard-step--top-container'}>
          <div className={'wizard-step--sparkle-container'}>
            <SparkleSpinner loading={status} />
          </div>
        </div>
        <div className={'wizard-step--footer'}>
          <div className={this.footerClass}>{this.footerText}</div>
        </div>
        <br />
      </>
    )
  }

  private get footerClass(): string {
    const {status} = this.props

    return classnames(`wizard-step--text-state`, {
      loading: status === RemoteDataState.Loading,
      success: status === RemoteDataState.Done,
      error: status === RemoteDataState.Error,
    })
  }

  private get footerText(): string {
    switch (this.props.status) {
      case RemoteDataState.Loading:
        return 'Loading...'
      case RemoteDataState.Done:
        return 'Data Written Successfully!'
      case RemoteDataState.Error:
        return 'Unable to Write Data'
    }
  }
}

export default StatusIndicator
