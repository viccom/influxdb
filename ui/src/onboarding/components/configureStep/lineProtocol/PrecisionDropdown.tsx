// Libraries
import React, {PureComponent} from 'react'

// Components
import {Dropdown} from 'src/clockface'

// Types
import {WritePrecision} from 'src/api'
import {Precision} from 'src/types/v2/dataLoaders'

interface Props {
  setPrecision: (precision: WritePrecision) => void
  precision: WritePrecision
}

const writePrecisions: WritePrecision[] = [
  WritePrecision.Ns,
  WritePrecision.Us,
  WritePrecision.Ms,
  WritePrecision.S,
]

const makePrecisionReadable = {
  [WritePrecision.Ns]: Precision.Nanoseconds,
  [WritePrecision.Us]: Precision.Microseconds,
  [WritePrecision.S]: Precision.Seconds,
  [WritePrecision.Ms]: Precision.Milliseconds,
}

class PrecisionDropdown extends PureComponent<Props> {
  public render() {
    const {setPrecision, precision} = this.props
    return (
      <div className={'wizard-step--footer dropdown'}>
        <label>Time Precision </label>
        <Dropdown
          selectedID={precision}
          onChange={setPrecision}
          widthPixels={200}
        >
          {writePrecisions.map(value => (
            <Dropdown.Item key={value} value={value} id={value}>
              {makePrecisionReadable[value]}
            </Dropdown.Item>
          ))}
        </Dropdown>
      </div>
    )
  }
}

export default PrecisionDropdown
