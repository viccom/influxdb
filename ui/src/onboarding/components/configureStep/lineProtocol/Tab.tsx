import React, {PureComponent} from 'react'

import {Radio} from 'src/clockface'

import {LineProtocolTab} from 'src/types/v2/dataLoaders'

interface Props {
  active: boolean
  tab: LineProtocolTab
  onClick: (tab: LineProtocolTab) => void
}

export default class extends PureComponent<Props> {
  public render() {
    const {tab, active} = this.props

    return (
      <Radio.Button
        key={tab}
        id={tab}
        titleText={tab}
        value={tab}
        active={active}
        onClick={this.handleClick}
      >
        {tab}
      </Radio.Button>
    )
  }

  private handleClick = () => {
    this.props.onClick(this.props.tab)
  }
}
