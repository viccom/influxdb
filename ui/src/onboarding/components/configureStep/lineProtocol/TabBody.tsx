import React, {PureComponent, ChangeEvent} from 'react'

import {
  Input,
  InputType,
  Form,
  Grid,
  Columns,
  ComponentSize,
} from 'src/clockface'
import DragAndDrop from 'src/shared/components/DragAndDrop'
import TextArea from 'src/clockface/components/inputs/TextArea'
import {LineProtocolTab} from 'src/types/v2/dataLoaders'
import {setLineProtocolBody} from 'src/onboarding/actions/dataLoaders'
import {WritePrecision} from 'src/api'

interface Props {
  lineProtocolBody: string
  activeLPTab: LineProtocolTab
  precision: WritePrecision
  setLineProtocolBody: typeof setLineProtocolBody
  onURLChange: (url: string) => void
  urlInput: string
  handleSubmit?: () => void
}

export default class extends PureComponent<Props> {
  public render() {
    const {lineProtocolBody, activeLPTab, urlInput} = this.props

    switch (activeLPTab) {
      case LineProtocolTab.UploadFile:
        return (
          <DragAndDrop
            submitText="Upload File"
            handleSubmit={this.handleSetLineProtocol}
            submitOnDrop={true}
            submitOnUpload={true}
          />
        )
      case LineProtocolTab.EnterManually:
        return (
          <TextArea
            value={lineProtocolBody}
            placeholder="Write text here"
            onChange={this.handleSetLineProtocol}
          />
        )
      case LineProtocolTab.EnterURL:
        return (
          <Grid>
            <Grid.Row>
              <Grid.Column
                widthXS={Columns.Twelve}
                widthMD={Columns.Ten}
                offsetMD={Columns.One}
              >
                <Form.Element label="File URL:">
                  <Input
                    titleText="File URL:"
                    type={InputType.Text}
                    placeholder="http://..."
                    value={urlInput}
                    onChange={this.handleChange}
                    autoFocus={true}
                    size={ComponentSize.Large}
                  />
                </Form.Element>
              </Grid.Column>
            </Grid.Row>
          </Grid>
        )
    }
  }

  private handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    const {value} = e.target
    this.props.onURLChange(value)
  }

  private handleSetLineProtocol = async (lpBody: string) => {
    const {setLineProtocolBody, handleSubmit} = this.props
    await setLineProtocolBody(lpBody)
    if (handleSubmit) {
      handleSubmit()
    }
  }
}
