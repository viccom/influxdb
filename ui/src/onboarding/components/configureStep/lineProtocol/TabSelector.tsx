import React, {PureComponent} from 'react'

import {Radio, ButtonShape} from 'src/clockface'
import {LineProtocolTab} from 'src/types/v2/dataLoaders'

import Tab from 'src/onboarding/components/configureStep/lineProtocol/Tab'

interface Props {
  tabs: LineProtocolTab[]
  activeLPTab: LineProtocolTab
  onClick: (tab: LineProtocolTab) => void
}

export default class extends PureComponent<Props> {
  public render() {
    const {tabs, activeLPTab} = this.props
    return (
      <Radio shape={ButtonShape.Default}>
        {tabs.map(t => (
          <Tab
            tab={t}
            key={t}
            active={activeLPTab === t}
            onClick={this.handleTabClick}
          />
        ))}
      </Radio>
    )
  }

  private handleTabClick = (tab: LineProtocolTab) => {
    const {activeLPTab, onClick} = this.props
    if (tab !== activeLPTab) {
      onClick(tab)
    }
  }
}
