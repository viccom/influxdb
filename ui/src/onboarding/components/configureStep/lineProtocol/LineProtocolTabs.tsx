// Libraries
import React, {PureComponent} from 'react'
import {connect} from 'react-redux'
import _ from 'lodash'

import PrecisionDropdown from 'src/onboarding/components/configureStep/lineProtocol/PrecisionDropdown'
import TabSelector from 'src/onboarding/components/configureStep/lineProtocol/TabSelector'
import TabBody from 'src/onboarding/components/configureStep/lineProtocol/TabBody'

// Types
import {LineProtocolTab} from 'src/types/v2/dataLoaders'

// Components
import {Grid, Columns} from 'src/clockface'

// Actions
import {
  setLineProtocolBody,
  setActiveLPTab,
  writeLineProtocolAction,
  setPrecision,
} from 'src/onboarding/actions/dataLoaders'

import {AppState} from 'src/types/v2/index'

// Styles
import 'src/clockface/components/auto_input/AutoInput.scss'
import {WritePrecision} from 'src/api'

interface OwnProps {
  tabs: LineProtocolTab[]
  bucket: string
  org: string
  handleSubmit?: () => void
}

type Props = OwnProps & DispatchProps & StateProps

interface DispatchProps {
  setLineProtocolBody: typeof setLineProtocolBody
  setActiveLPTab: typeof setActiveLPTab
  writeLineProtocolAction: typeof writeLineProtocolAction
  setPrecision: typeof setPrecision
}

interface StateProps {
  lineProtocolBody: string
  activeLPTab: LineProtocolTab
  precision: WritePrecision
}

interface State {
  urlInput: string
}

export class LineProtocolTabs extends PureComponent<Props, State> {
  constructor(props: Props) {
    super(props)

    this.state = {
      urlInput: '',
    }
  }

  public render() {
    const {
      setPrecision,
      precision,
      activeLPTab,
      tabs,
      setLineProtocolBody,
      lineProtocolBody,
      handleSubmit,
    } = this.props

    const {urlInput} = this.state

    return (
      <div>
        <TabSelector
          activeLPTab={activeLPTab}
          tabs={tabs}
          onClick={this.handleTabClick}
        />

        <Grid>
          <Grid.Row>
            <Grid.Column
              widthXS={Columns.Twelve}
              widthMD={Columns.Ten}
              offsetMD={Columns.One}
              widthLG={Columns.Eight}
              offsetLG={Columns.Two}
            >
              <div className="onboarding--admin-user-form">
                <div className={'wizard-step--lp-body'}>
                  <TabBody
                    onURLChange={this.handleURLChange}
                    activeLPTab={activeLPTab}
                    precision={precision}
                    urlInput={urlInput}
                    lineProtocolBody={lineProtocolBody}
                    setLineProtocolBody={setLineProtocolBody}
                    handleSubmit={handleSubmit}
                  />
                </div>

                <PrecisionDropdown
                  setPrecision={setPrecision}
                  precision={precision}
                />
              </div>
            </Grid.Column>
          </Grid.Row>
        </Grid>
      </div>
    )
  }

  private handleTabClick = (tab: LineProtocolTab) => {
    const {setActiveLPTab, setLineProtocolBody} = this.props

    setLineProtocolBody('')
    setActiveLPTab(tab)
  }

  private handleURLChange = (urlInput: string) => {
    this.setState({urlInput})
  }
}

const mstp = ({
  dataLoading: {
    dataLoaders: {lineProtocolBody, activeLPTab, precision},
  },
}: AppState) => {
  return {lineProtocolBody, activeLPTab, precision}
}

const mdtp: DispatchProps = {
  setLineProtocolBody,
  setActiveLPTab,
  writeLineProtocolAction,
  setPrecision,
}

export default connect<StateProps, DispatchProps, OwnProps>(
  mstp,
  mdtp
)(LineProtocolTabs)
