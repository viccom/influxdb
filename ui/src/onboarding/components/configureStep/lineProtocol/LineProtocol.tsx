// Libraries
import React, {PureComponent} from 'react'
import {connect} from 'react-redux'

import _ from 'lodash'

// Components
import LineProtocolTabs from 'src/onboarding/components/configureStep/lineProtocol/LineProtocolTabs'
import OnboardingButtons from 'src/onboarding/components/OnboardingButtons'
import {Form} from 'src/clockface'
import FancyScrollbar from 'src/shared/components/fancy_scrollbar/FancyScrollbar'

// Actions
import {
  setLPStatus as setLPStatusAction,
  writeLineProtocolAction,
} from 'src/onboarding/actions/dataLoaders'

// Decorator
import {ErrorHandling} from 'src/shared/decorators/errors'

// Types
import {LineProtocolTab} from 'src/types/v2/dataLoaders'
import {AppState} from 'src/types/v2/index'
import {WritePrecision} from 'src/api'
import {RemoteDataState} from 'src/types'

interface OwnProps {
  bucket: string
  org: string
  onClickNext: () => void
  onClickBack: () => void
  onClickSkip: () => void
}

interface StateProps {
  lineProtocolBody: string
  precision: WritePrecision
}

interface DispatchProps {
  setLPStatus: typeof setLPStatusAction
  writeLineProtocolAction: typeof writeLineProtocolAction
}

type Props = OwnProps & StateProps & DispatchProps

@ErrorHandling
export class LineProtocol extends PureComponent<Props> {
  public componentDidMount() {
    const {setLPStatus} = this.props
    setLPStatus(RemoteDataState.NotStarted)
  }

  public render() {
    return (
      <Form onSubmit={this.handleSubmit}>
        <div className="wizard-step--scroll-area">
          <FancyScrollbar autoHide={false}>
            <div className="wizard-step--scroll-content">
              <h3 className="wizard-step--title">Add Data via Line Protocol</h3>
              <h5 className="wizard-step--sub-title">
                Need help writing InfluxDB Line Protocol? See Documentation
              </h5>
              {this.Content}
            </div>
          </FancyScrollbar>
        </div>
        <OnboardingButtons
          onClickBack={this.props.onClickBack}
          onClickSkip={this.props.onClickSkip}
          showSkip={true}
          autoFocusNext={true}
          skipButtonText={'Skip Config'}
        />
      </Form>
    )
  }

  private get LineProtocolTabs(): LineProtocolTab[] {
    return [LineProtocolTab.UploadFile, LineProtocolTab.EnterManually]
  }

  private get Content(): JSX.Element {
    const {bucket, org} = this.props
    return (
      <LineProtocolTabs
        tabs={this.LineProtocolTabs}
        bucket={bucket}
        org={org}
        handleSubmit={this.handleSubmit}
      />
    )
  }

  private handleSubmit = () => {
    const {
      bucket,
      org,
      writeLineProtocolAction,
      lineProtocolBody,
      precision,
    } = this.props

    writeLineProtocolAction(org, bucket, lineProtocolBody, precision)
    this.props.onClickNext()
  }
}

const mstp = ({
  dataLoading: {
    dataLoaders: {lineProtocolBody, precision},
  },
}: AppState): StateProps => {
  return {lineProtocolBody, precision}
}

const mdtp: DispatchProps = {
  setLPStatus: setLPStatusAction,
  writeLineProtocolAction,
}

export default connect<StateProps, DispatchProps, OwnProps>(
  mstp,
  mdtp
)(LineProtocol)
