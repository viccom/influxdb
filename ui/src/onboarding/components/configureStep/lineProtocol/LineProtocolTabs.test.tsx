// Libraries
import React from 'react'
import {shallow} from 'enzyme'

// Components
import {LineProtocolTabs} from 'src/onboarding/components/configureStep/lineProtocol/LineProtocolTabs'

import {LineProtocolTab} from 'src/types/v2/dataLoaders'

const setup = (override?) => {
  const props = {
    tabs: [
      LineProtocolTab.UploadFile,
      LineProtocolTab.EnterManually,
      LineProtocolTab.EnterURL,
    ],
    bucket: 'a',
    org: 'a',
    ...override,
  }

  const wrapper = shallow(<LineProtocolTabs {...props} />)

  return {wrapper}
}

describe('LineProtocolTabs', () => {
  describe('rendering', () => {
    it('renders!', () => {
      const {wrapper} = setup()
      expect(wrapper.exists()).toBe(true)

      expect(wrapper).toMatchSnapshot()
    })
  })
})
