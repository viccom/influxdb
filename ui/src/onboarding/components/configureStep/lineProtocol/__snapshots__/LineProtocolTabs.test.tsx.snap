// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineProtocolTabs rendering renders! 1`] = `
<div>
  <default_1
    onClick={[Function]}
    tabs={
      Array [
        "Upload File",
        "Enter Manually",
        "Enter URL",
      ]
    }
  />
  <Grid>
    <GridRow>
      <GridColumn
        offsetLG={2}
        offsetMD={1}
        widthLG={8}
        widthMD={10}
        widthXS={12}
      >
        <div
          className="onboarding--admin-user-form"
        >
          <div
            className="wizard-step--lp-body"
          >
            <default_1
              onURLChange={[Function]}
              urlInput=""
            />
          </div>
          <PrecisionDropdown />
        </div>
      </GridColumn>
    </GridRow>
  </Grid>
</div>
`;
